File: Sample/Global/FocusManager.py
# -*- encoding: utf-8 -*-
r"""\brief Keyboard focus manager for runtime UI navigation."""

from __future__ import annotations
from enum import Enum
from typing import Dict, List, Optional, Set, Union
from Engine.UI.Base import ControlBase, FunctionalBase
from Engine.UI.Base.FocusableMixin import Direction


class FocusTransition(Enum):
    r"""\brief Strategy used when moving from one focus group to a neighbour."""

    DIRECTIONAL = "directional"
    EXPLICIT = "explicit"


class FocusNeighbor:
    r"""\brief Directional neighbour metadata for a focus group."""

    def __init__(
        self,
        group: "FocusGroup",
        transition: FocusTransition = FocusTransition.DIRECTIONAL,
    ) -> None:
        r"""\brief Construct a focus neighbour edge.

        - \param group Target focus group.
        - \param transition Navigation strategy for this edge.
        """
        self.group = group
        self.transition = transition


class FocusGroup:
    r"""\brief Group of controls that share a directional focus neighbourhood."""

    def __init__(
        self,
        name: str,
        items: Optional[List[FunctionalBase]] = None,
        activeOwner: Optional[FunctionalBase] = None,
    ) -> None:
        r"""\brief Construct a focus group.

        - \param name Debug name for the group.
        - \param items Initial focusable controls in the group.
        - \param activeOwner Control whose active state gates group entry.
        """
        self.name = name
        self.activeOwner = activeOwner
        self._items: List[FunctionalBase] = []
        self._neighborMap: Dict[Direction, FocusNeighbor] = {}
        self._lastFocusedElement: Optional[FunctionalBase] = None
        if items is not None:
            for item in items:
                self.addItem(item)

    def addItem(self, item: FunctionalBase) -> None:
        r"""\brief Add a control to this focus group.

        - \param item Control to add.
        """
        if item in self._items:
            return
        self._items.append(item)
        item.setFocusGroup(self)

    def removeItem(self, item: FunctionalBase) -> None:
        r"""\brief Remove a control from this focus group.

        - \param item Control to remove.
        """
        if item not in self._items:
            return
        self._items.remove(item)
        if item.getFocusGroup() is self:
            item.setFocusGroup(None)
        if self._lastFocusedElement is item:
            self._lastFocusedElement = None

    def getItems(self) -> List[FunctionalBase]:
        r"""\brief Return controls registered in this group.

        - \return Registered focusable controls.
        """
        return self._items[:]

    def setNeighbor(
        self,
        direction: Direction,
        neighbor: Union["FocusGroup", FocusNeighbor],
        transition: FocusTransition = FocusTransition.DIRECTIONAL,
    ) -> None:
        r"""\brief Set the neighbouring group for a direction.

        - \param direction Navigation direction.
        - \param neighbor Target group or neighbour metadata.
        - \param transition Navigation strategy when `neighbor` is a group.
        """
        if isinstance(neighbor, FocusNeighbor):
            self._neighborMap[direction] = neighbor
        else:
            self._neighborMap[direction] = FocusNeighbor(neighbor, transition)

    def getNeighbor(self, direction: Direction) -> Optional[FocusNeighbor]:
        r"""\brief Return the neighbour for a direction.

        - \param direction Navigation direction.
        - \return Focus neighbour or None.
        """
        return self._neighborMap.get(direction)

    def canEnter(self) -> bool:
        r"""\brief Return whether focus may enter this group.

        - \return True if the group has an active focusable target.
        """
        if self.activeOwner is not None and not self._isOwnerAvailable(self.activeOwner):
            return False
        return self._findInitialFocusLocal() is not None

    def findInitialFocus(self) -> Optional[FunctionalBase]:
        r"""\brief Return the focus target to use when entering this group.

        - \return Focusable control or None.
        """
        if self.activeOwner is not None and not self._isOwnerAvailable(self.activeOwner):
            return None
        return self._findInitialFocusLocal()

    def rememberFocus(self, element: FunctionalBase) -> None:
        r"""\brief Store the last focused element in this group.

        - \param element Focused element.
        """
        if element in self._items:
            self._lastFocusedElement = element

    def moveWithin(self, current: FunctionalBase, direction: Direction) -> Optional[FunctionalBase]:
        r"""\brief Move within the group before trying neighbour groups.

        - \param current Currently focused element.
        - \param direction Navigation direction.
        - \return New focus target or None when the group does not handle it.
        """
        return None

    def _findInitialFocusLocal(self) -> Optional[FunctionalBase]:
        if self._lastFocusedElement is not None and self._isLocallyFocusable(self._lastFocusedElement):
            return self._lastFocusedElement
        for item in self._items:
            if self._isLocallyFocusable(item):
                return item
        return None

    @staticmethod
    def _isLocallyFocusable(element: FunctionalBase) -> bool:
        return element.canReceiveFocus()

    @staticmethod
    def _isOwnerAvailable(element: FunctionalBase) -> bool:
        if not element.getActive():
            return False
        if isinstance(element, ControlBase) and not element.getVisible():
            return False
        return True


class FocusManager:
    r"""\brief Scene-level keyboard focus manager."""

    def __init__(self) -> None:
        r"""\brief Construct a focus manager with navigation disabled."""
        self._focusedElement: Optional[FunctionalBase] = None
        self._cursorFocusElement: Optional[FunctionalBase] = None
        self._focusGroups: List[FocusGroup] = []
        self._autoFocusGroups: Dict[FunctionalBase, FocusGroup] = {}
        self._navigationEnabled: bool = False

    def setNavigationEnabled(self, enabled: bool) -> None:
        r"""\brief Enable or disable focus-based keyboard routing.

        - \param enabled True to route keyboard events only to the focused control.
        """
        self._navigationEnabled = enabled
        if not enabled:
            self.clearFocus()

    def getNavigationEnabled(self) -> bool:
        r"""\brief Return whether focus routing is enabled.

        - \return True when focus routing is enabled.
        """
        return self._navigationEnabled

    def isRoutingKeyboard(self) -> bool:
        r"""\brief Return whether keyboard input should be focus-routed.

        - \return True when focus routing is enabled.
        """
        return self._navigationEnabled

    def registerElement(self, element: FunctionalBase) -> None:
        r"""\brief Register a top-level focusable element for automatic focus.

        - \param element UI element to register.
        """
        for group in self._focusGroups:
            if element in group.getItems():
                return
        if element in self._autoFocusGroups:
            return
        name = element.getName() if isinstance(element, ControlBase) else ""
        group = FocusGroup(f"auto:{name or id(element)}", [element], element)
        self._autoFocusGroups[element] = group

    def unregisterElement(self, element: FunctionalBase) -> None:
        r"""\brief Unregister a top-level focusable element.

        - \param element UI element to unregister.
        """
        group = self._autoFocusGroups.pop(element, None)
        if group is not None:
            group.removeItem(element)
        if self._focusedElement is element:
            self.clearFocus()
        if self._cursorFocusElement is element:
            self._cursorFocusElement = None

    def registerFocusGroup(self, group: FocusGroup) -> None:
        r"""\brief Register an explicit focus group.

        - \param group Focus group to register.
        """
        if group not in self._focusGroups:
            self._focusGroups.append(group)

    def unregisterFocusGroup(self, group: FocusGroup) -> None:
        r"""\brief Unregister an explicit focus group.

        - \param group Focus group to unregister.
        """
        if group in self._focusGroups:
            self._focusGroups.remove(group)
        if self._focusedElement is not None and self._findGroupForElement(self._focusedElement) is group:
            self.clearFocus()

    def getFocus(self) -> Optional[FunctionalBase]:
        r"""\brief Return the currently focused element.

        - \return Focused element or None.
        """
        return self._focusedElement

    def setFocus(self, element: FunctionalBase) -> bool:
        r"""\brief Move keyboard focus to an element.

        - \param element Target element.
        - \return True if focus moved to the target.
        """
        if not self._isFocusable(element):
            return False
        if self._focusedElement is element:
            self._cursorFocusElement = element
            return True
        self.clearFocus()
        self._focusedElement = element
        self._cursorFocusElement = element
        element.setFocused(True)
        group = self._findGroupForElement(element)
        if group is not None:
            group.rememberFocus(element)
        return True

    def clearFocus(self) -> None:
        r"""\brief Clear the current keyboard focus."""
        if self._focusedElement is None:
            return
        focusedElement = self._focusedElement
        self._focusedElement = None
        if self._cursorFocusElement is focusedElement:
            self._cursorFocusElement = None
        focusedElement.setFocused(False)

    def prepareFrame(self) -> None:
        r"""\brief Validate focus before UI update dispatch for the frame."""
        if not self._navigationEnabled:
            return
        if self._focusedElement is not None and self._isFocusable(self._focusedElement):
            return
        self.clearFocus()
        nextFocus = self._findDefaultFocus()
        if nextFocus is not None:
            self.setFocus(nextFocus)

    def shouldDispatchKeyboardTo(self, element: FunctionalBase) -> bool:
        r"""\brief Return whether an element may receive keyboard callbacks.

        - \param element UI element being updated.
        - \return True if keyboard callbacks should be dispatched to it.
        """
        if not self._navigationEnabled:
            return True
        return self._focusedElement is element and self._isFocusable(element)

    def isFocused(self, element: FunctionalBase) -> bool:
        r"""\brief Return whether an element is focused.

        - \param element UI element to test.
        - \return True if this element owns focus.
        """
        return self._focusedElement is element

    def isCursorFocusOwner(self, element: FunctionalBase) -> bool:
        r"""\brief Return whether an element owns the selection cursor.

        - \param element UI element to test.
        - \return True if this element owns the keyboard selection cursor.
        """
        if not self._navigationEnabled:
            return False
        return self._cursorFocusElement is element and self._isFocusable(element)

    def requestDirectionalMove(self, element: FunctionalBase, direction: Direction) -> bool:
        r"""\brief Request a directional focus move from an element.

        - \param element Source element.
        - \param direction Navigation direction.
        - \return True if focus moved.
        """
        if not self._navigationEnabled:
            return False
        if self._focusedElement is not element:
            return False
        return self.moveFocus(direction, element)

    def moveFocus(self, direction: Direction, source: Optional[FunctionalBase] = None) -> bool:
        r"""\brief Move focus in a direction.

        - \param direction Navigation direction.
        - \param source Optional source element; current focus is used when omitted.
        - \return True if focus moved.
        """
        if not self._navigationEnabled:
            return False
        current = source if source is not None else self._focusedElement
        if current is None:
            return False
        group = self._findGroupForElement(current)
        if group is None:
            return False
        withinTarget = group.moveWithin(current, direction)
        if withinTarget is not None:
            return self.setFocus(withinTarget)
        neighborTarget = self._findDirectionalTarget(group, direction)
        if neighborTarget is None:
            return False
        return self.setFocus(neighborTarget)

    def activateGroup(self, group: FocusGroup) -> bool:
        r"""\brief Focus the default element in a group.

        - \param group Focus group to activate.
        - \return True if focus moved into the group.
        """
        target = group.findInitialFocus()
        if target is None:
            return False
        return self.setFocus(target)

    def _findDefaultFocus(self) -> Optional[FunctionalBase]:
        for group in self._allGroups():
            target = group.findInitialFocus()
            if target is not None:
                return target
        return None

    def _findDirectionalTarget(self, group: FocusGroup, direction: Direction) -> Optional[FunctionalBase]:
        currentGroup = group
        visited: Set[int] = {id(currentGroup)}
        while True:
            neighbor = currentGroup.getNeighbor(direction)
            if neighbor is None or neighbor.transition != FocusTransition.DIRECTIONAL:
                return None
            nextGroup = neighbor.group
            if id(nextGroup) in visited:
                return None
            visited.add(id(nextGroup))
            target = nextGroup.findInitialFocus()
            if target is not None:
                return target
            currentGroup = nextGroup

    def _isFocusable(self, element: FunctionalBase) -> bool:
        if not element.canReceiveFocus():
            return False
        group = self._findGroupForElement(element)
        if group is not None and not group.canEnter():
            return False
        return True

    def _findGroupForElement(self, element: FunctionalBase) -> Optional[FocusGroup]:
        for group in self._focusGroups:
            if element in group.getItems():
                return group
        explicitGroup = element.getFocusGroup()
        if isinstance(explicitGroup, FocusGroup) and explicitGroup in self._focusGroups:
            return explicitGroup
        return self._autoFocusGroups.get(element)

    def _allGroups(self) -> List[FocusGroup]:
        return self._focusGroups + list(self._autoFocusGroups.values())


__all__ = ["FocusManager", "FocusGroup", "FocusNeighbor", "FocusTransition"]

File: Sample/Core/Global/src/FocusManager.cpp
#include <FocusManager.hpp>

#include "FocusManager/FocusRuntime.hpp"

#include <UI/ControlBase.hpp>

#include <algorithm>
#include <sstream>
#include <unordered_set>
#include <utility>

const std::string FocusTransition::DIRECTIONAL = "directional";
const std::string FocusTransition::EXPLICIT = "explicit";

FocusNeighbor::FocusNeighbor(std::shared_ptr<FocusGroup> groupValue,
                             std::string transitionValue)
    : group_(std::move(groupValue)), transition(std::move(transitionValue)) {}

std::shared_ptr<FocusGroup> FocusNeighbor::getGroup() const {
    return group_.lock();
}

void FocusNeighbor::setGroup(const std::shared_ptr<FocusGroup>& group) {
    group_ = group;
}

FocusGroup::FocusGroup(std::string nameValue,
                       std::vector<std::shared_ptr<FunctionalBase>> items,
                       std::shared_ptr<FunctionalBase> activeOwnerValue)
    : name(std::move(nameValue)), activeOwner(std::move(activeOwnerValue)) {
    for (const std::shared_ptr<FunctionalBase>& item : items) {
        addItem(item);
    }
}

void FocusGroup::addItem(const std::shared_ptr<FunctionalBase>& item) {
    if (item == nullptr || contains(item)) {
        return;
    }
    items_.push_back(item);
    const std::shared_ptr<FocusGroup> self = self_.lock();
    if (self != nullptr) {
        item->setFocusGroup(self);
    }
}

void FocusGroup::removeItem(const std::shared_ptr<FunctionalBase>& item) {
    const auto iterator = std::find(items_.begin(), items_.end(), item);
    if (iterator == items_.end()) {
        return;
    }
    items_.erase(iterator);
    const std::shared_ptr<RuntimeObject> assignedGroup = item->getFocusGroup();
    if (assignedGroup.get() == this) {
        item->setFocusGroup(nullptr);
    }
    if (lastFocusedElement_.lock() == item) {
        lastFocusedElement_.reset();
    }
}

std::vector<std::shared_ptr<FunctionalBase>> FocusGroup::getItems() const {
    return items_;
}

void FocusGroup::setNeighbor(const std::string& direction,
                             const std::shared_ptr<FocusGroup>& neighbor,
                             const std::string& transition) {
    neighborMap_[direction] =
        std::make_shared<FocusNeighbor>(neighbor, transition);
}

void FocusGroup::setNeighbor(const std::string& direction,
                             const std::shared_ptr<FocusNeighbor>& neighbor) {
    neighborMap_[direction] = neighbor;
}

std::shared_ptr<FocusNeighbor> FocusGroup::getNeighbor(
    const std::string& direction) const {
    const auto iterator = neighborMap_.find(direction);
    return iterator == neighborMap_.end() ? nullptr : iterator->second;
}

bool FocusGroup::canEnter() const {
    if (activeOwner != nullptr && !isOwnerAvailable(activeOwner)) {
        return false;
    }
    return findInitialFocusLocal() != nullptr;
}

std::shared_ptr<FunctionalBase> FocusGroup::findInitialFocus() const {
    if (activeOwner != nullptr && !isOwnerAvailable(activeOwner)) {
        return nullptr;
    }
    return findInitialFocusLocal();
}

void FocusGroup::rememberFocus(const std::shared_ptr<FunctionalBase>& element) {
    if (contains(element)) {
        lastFocusedElement_ = element;
    }
}

std::shared_ptr<FunctionalBase> FocusGroup::moveWithin(
    const std::shared_ptr<FunctionalBase>& current,
    const std::string& direction) {
    static_cast<void>(current);
    static_cast<void>(direction);
    return nullptr;
}

void FocusGroup::releaseRuntimeState() noexcept {
    for (const std::shared_ptr<FunctionalBase>& item : items_) {
        if (item != nullptr && item->getFocusGroup().get() == this) {
            item->setFocusGroup(nullptr);
        }
    }
    activeOwner.reset();
    neighborMap_.clear();
    items_.clear();
    lastFocusedElement_.reset();
    self_.reset();
}

bool FocusGroup::contains(
    const std::shared_ptr<FunctionalBase>& element) const {
    return std::find(items_.begin(), items_.end(), element) != items_.end();
}

std::shared_ptr<FunctionalBase> FocusGroup::findInitialFocusLocal() const {
    const std::shared_ptr<FunctionalBase> previous = lastFocusedElement_.lock();
    if (isLocallyFocusable(previous)) {
        return previous;
    }
    const auto iterator =
        std::find_if(items_.begin(), items_.end(), isLocallyFocusable);
    return iterator == items_.end() ? nullptr : *iterator;
}

bool FocusGroup::isLocallyFocusable(
    const std::shared_ptr<FunctionalBase>& element) {
    return element != nullptr && element->canReceiveFocus();
}

bool FocusGroup::isOwnerAvailable(
    const std::shared_ptr<FunctionalBase>& element) {
    if (element == nullptr || !element->getActive()) {
        return false;
    }
    const ControlBase* control =
        dynamic_cast<const ControlBase*>(element.get());
    return control == nullptr || control->getVisible();
}

FocusManager::~FocusManager() {
    shutdown();
}

void FocusManager::shutdown() noexcept {
    navigationEnabled_ = false;
    focusedElement_.reset();
    cursorFocusElement_.reset();
    for (const auto& [element, group] : autoFocusGroups_) {
        static_cast<void>(element);
        if (group != nullptr) {
            group->releaseRuntimeState();
        }
    }
    autoFocusGroups_.clear();
    for (const std::shared_ptr<FocusGroup>& group : focusGroups_) {
        if (group != nullptr) {
            group->releaseRuntimeState();
        }
    }
    focusGroups_.clear();
}

void FocusManager::setNavigationEnabled(bool enabled) {
    navigationEnabled_ = enabled;
    if (!enabled) {
        clearFocus();
    }
}

bool FocusManager::getNavigationEnabled() const {
    return navigationEnabled_;
}

bool FocusManager::isRoutingKeyboard() const {
    return navigationEnabled_;
}

void FocusManager::registerElement(
    const std::shared_ptr<FunctionalBase>& element) {
    if (element == nullptr) {
        return;
    }
    for (const std::shared_ptr<FocusGroup>& group : focusGroups_) {
        if (group != nullptr && group->contains(element)) {
            return;
        }
    }
    if (autoFocusGroups_.contains(element.get())) {
        return;
    }
    std::string label;
    const ControlBase* control =
        dynamic_cast<const ControlBase*>(element.get());
    if (control != nullptr) {
        label = control->getName();
    }
    if (label.empty()) {
        std::ostringstream stream;
        stream << element.get();
        label = stream.str();
    }
    const std::shared_ptr<FocusGroup> group = std::make_shared<FocusGroup>(
        "auto:" + label, std::vector<std::shared_ptr<FunctionalBase>>{element},
        element);
    group->self_ = group;
    autoFocusGroups_[element.get()] = group;
    element->setFocusGroup(group);
}

void FocusManager::unregisterElement(
    const std::shared_ptr<FunctionalBase>& element) {
    if (element == nullptr) {
        return;
    }
    const auto iterator = autoFocusGroups_.find(element.get());
    if (iterator != autoFocusGroups_.end()) {
        iterator->second->removeItem(element);
        autoFocusGroups_.erase(iterator);
    }
    if (focusedElement_ == element) {
        clearFocus();
    }
    if (cursorFocusElement_ == element) {
        cursorFocusElement_.reset();
    }
}

void FocusManager::registerFocusGroup(
    const std::shared_ptr<FocusGroup>& group) {
    if (group != nullptr && std::find(focusGroups_.begin(), focusGroups_.end(),
                                      group) == focusGroups_.end()) {
        group->self_ = group;
        focusGroups_.push_back(group);
        for (const std::shared_ptr<FunctionalBase>& item : group->getItems()) {
            item->setFocusGroup(group);
        }
    }
}

void FocusManager::unregisterFocusGroup(
    const std::shared_ptr<FocusGroup>& group) {
    focusGroups_.erase(
        std::remove(focusGroups_.begin(), focusGroups_.end(), group),
        focusGroups_.end());
    if (focusedElement_ != nullptr &&
        findGroupForElement(focusedElement_) == group) {
        clearFocus();
    }
}

std::shared_ptr<FunctionalBase> FocusManager::getFocus() const {
    return focusedElement_;
}

bool FocusManager::setFocus(const std::shared_ptr<FunctionalBase>& element) {
    if (!isFocusable(element)) {
        return false;
    }
    if (focusedElement_ == element) {
        cursorFocusElement_ = element;
        return true;
    }
    clearFocus();
    focusedElement_ = element;
    cursorFocusElement_ = element;
    element->setFocused(true);
    const std::shared_ptr<FocusGroup> group = findGroupForElement(element);
    if (group != nullptr) {
        group->rememberFocus(element);
    }
    return true;
}

void FocusManager::clearFocus() {
    if (focusedElement_ == nullptr) {
        return;
    }
    const std::shared_ptr<FunctionalBase> previous = focusedElement_;
    focusedElement_.reset();
    if (cursorFocusElement_ == previous) {
        cursorFocusElement_.reset();
    }
    previous->setFocused(false);
}

void FocusManager::prepareFrame() {
    if (!navigationEnabled_) {
        return;
    }
    if (focusedElement_ != nullptr && isFocusable(focusedElement_)) {
        return;
    }
    clearFocus();
    const std::shared_ptr<FunctionalBase> next = findDefaultFocus();
    if (next != nullptr) {
        setFocus(next);
    }
}

bool FocusManager::shouldDispatchKeyboardTo(
    const std::shared_ptr<FunctionalBase>& element) const {
    return !navigationEnabled_ ||
           (focusedElement_ == element && isFocusable(element));
}

bool FocusManager::isFocused(
    const std::shared_ptr<FunctionalBase>& element) const {
    return focusedElement_ == element;
}

bool FocusManager::isCursorFocusOwner(
    const std::shared_ptr<FunctionalBase>& element) const {
    return navigationEnabled_ && cursorFocusElement_ == element &&
           isFocusable(element);
}

bool FocusManager::requestDirectionalMove(
    const std::shared_ptr<FunctionalBase>& element,
    const std::string& direction) {
    return navigationEnabled_ && focusedElement_ == element &&
           moveFocus(direction, element);
}

bool FocusManager::moveFocus(const std::string& direction,
                             const std::shared_ptr<FunctionalBase>& source) {
    if (!navigationEnabled_) {
        return false;
    }
    const std::shared_ptr<FunctionalBase> current =
        source == nullptr ? focusedElement_ : source;
    if (current == nullptr) {
        return false;
    }
    const std::shared_ptr<FocusGroup> group = findGroupForElement(current);
    if (group == nullptr) {
        return false;
    }
    const std::shared_ptr<FunctionalBase> within =
        group->moveWithin(current, direction);
    if (within != nullptr) {
        return setFocus(within);
    }
    const std::shared_ptr<FunctionalBase> target =
        findDirectionalTarget(group, direction);
    return target != nullptr && setFocus(target);
}

bool FocusManager::activateGroup(const std::shared_ptr<FocusGroup>& group) {
    if (group == nullptr) {
        return false;
    }
    const std::shared_ptr<FunctionalBase> target = group->findInitialFocus();
    return target != nullptr && setFocus(target);
}

bool FocusManager::setFocus(FunctionalBase& element) {
    const std::shared_ptr<FunctionalBase> target = findElement(&element);
    return target != nullptr && setFocus(target);
}

bool FocusManager::shouldDispatchKeyboardTo(
    const FunctionalBase& element) const {
    const std::shared_ptr<FunctionalBase> target = findElement(&element);
    return !navigationEnabled_ ||
           (target != nullptr && shouldDispatchKeyboardTo(target));
}

bool FocusManager::isCursorFocusOwner(const FunctionalBase& element) const {
    const std::shared_ptr<FunctionalBase> target = findElement(&element);
    return target != nullptr && isCursorFocusOwner(target);
}

bool FocusManager::requestDirectionalMove(FunctionalBase& element,
                                          const std::string& direction) {
    const std::shared_ptr<FunctionalBase> source = findElement(&element);
    return source != nullptr && requestDirectionalMove(source, direction);
}

std::shared_ptr<FunctionalBase> FocusManager::findDefaultFocus() const {
    for (const std::shared_ptr<FocusGroup>& group : allGroups()) {
        if (group == nullptr) {
            continue;
        }
        const std::shared_ptr<FunctionalBase> target =
            group->findInitialFocus();
        if (target != nullptr) {
            return target;
        }
    }
    return nullptr;
}

std::shared_ptr<FunctionalBase> FocusManager::findDirectionalTarget(
    const std::shared_ptr<FocusGroup>& group,
    const std::string& direction) const {
    using ludork::global::focus_manager_impl::ElementId;
    using ludork::global::focus_manager_impl::FocusRuntime;
    using ludork::global::focus_manager_impl::GroupId;
    using ludork::global::focus_manager_impl::GroupState;
    using ludork::global::focus_manager_impl::NeighborState;

    const std::vector<std::shared_ptr<FocusGroup>> groups = allGroups();
    std::unordered_map<GroupId, std::shared_ptr<FocusGroup>> publicGroups;
    FocusRuntime runtime;
    for (const std::shared_ptr<FocusGroup>& publicGroup : groups) {
        if (publicGroup == nullptr) {
            continue;
        }
        const GroupId id = reinterpret_cast<GroupId>(publicGroup.get());
        publicGroups.emplace(id, publicGroup);
        GroupState state;
        state.activeOwner =
            reinterpret_cast<ElementId>(publicGroup->activeOwner.get());
        for (const std::shared_ptr<FunctionalBase>& item :
             publicGroup->getItems()) {
            state.elements.push_back(reinterpret_cast<ElementId>(item.get()));
        }
        const std::shared_ptr<FocusNeighbor> neighbor =
            publicGroup->getNeighbor(direction);
        const std::shared_ptr<FocusGroup> target =
            neighbor == nullptr ? nullptr : neighbor->getGroup();
        if (neighbor != nullptr) {
            state.neighbors.emplace(
                direction,
                NeighborState{reinterpret_cast<GroupId>(target.get()),
                              neighbor->transition});
        }
        runtime.addGroup(id, std::move(state));
    }
    const GroupId start = reinterpret_cast<GroupId>(group.get());
    for (const GroupId id : runtime.directionalPath(start, direction)) {
        const auto iterator = publicGroups.find(id);
        if (iterator == publicGroups.end()) {
            return nullptr;
        }
        const std::shared_ptr<FunctionalBase> target =
            iterator->second->findInitialFocus();
        if (target != nullptr) {
            return target;
        }
    }
    return nullptr;
}

bool FocusManager::isFocusable(
    const std::shared_ptr<FunctionalBase>& element) const {
    if (element == nullptr || !element->canReceiveFocus()) {
        return false;
    }
    const std::shared_ptr<FocusGroup> group = findGroupForElement(element);
    return group == nullptr || group->canEnter();
}

std::shared_ptr<FocusGroup> FocusManager::findGroupForElement(
    const std::shared_ptr<FunctionalBase>& element) const {
    if (element == nullptr) {
        return nullptr;
    }
    for (const std::shared_ptr<FocusGroup>& group : focusGroups_) {
        if (group != nullptr && group->contains(element)) {
            return group;
        }
    }
    const auto iterator = autoFocusGroups_.find(element.get());
    return iterator == autoFocusGroups_.end() ? nullptr : iterator->second;
}

std::vector<std::shared_ptr<FocusGroup>> FocusManager::allGroups() const {
    std::vector<std::shared_ptr<FocusGroup>> result = focusGroups_;
    result.reserve(result.size() + autoFocusGroups_.size());
    for (const auto& [element, group] : autoFocusGroups_) {
        static_cast<void>(element);
        result.push_back(group);
    }
    return result;
}

std::shared_ptr<FunctionalBase> FocusManager::findElement(
    const FunctionalBase* element) const {
    if (element == nullptr) {
        return nullptr;
    }
    if (focusedElement_.get() == element) {
        return focusedElement_;
    }
    if (cursorFocusElement_.get() == element) {
        return cursorFocusElement_;
    }
    for (const std::shared_ptr<FocusGroup>& group : allGroups()) {
        if (group == nullptr) {
            continue;
        }
        for (const std::shared_ptr<FunctionalBase>& item : group->getItems()) {
            if (item.get() == element) {
                return item;
            }
        }
    }
    const ControlBase* control = dynamic_cast<const ControlBase*>(element);
    if (control == nullptr) {
        return nullptr;
    }
    const std::shared_ptr<const ControlBase> owner =
        control->weak_from_this().lock();
    const std::shared_ptr<const FunctionalBase> functional =
        std::dynamic_pointer_cast<const FunctionalBase>(owner);
    return std::const_pointer_cast<FunctionalBase>(functional);
}
